cmd/ctr/commands/images/mount.go

```diff
@@ -116,23 +117,46 @@ When you are done, use the unmount command.
 		s := client.SnapshotService(snapshotter)
 
 		var mounts []mount.Mount
-		if context.Bool("rw") {
+
+		rw := context.Bool("rw")
+		if rw {
 			mounts, err = s.Prepare(ctx, target, chainID)
 		} else {
 			mounts, err = s.View(ctx, target, chainID)
 		}
+		shouldCleanup := true
 		if err != nil {
 			if errdefs.IsAlreadyExists(err) {
+				shouldCleanup = false
+
+				info, err2 := s.Stat(ctx, target)
+				if err2 != nil {
+					return errors.Wrap(err2, "error stating already existing snapshot")
+				}
+
+				if rw && info.Kind == snapshots.KindView {
+					return errors.New("cannot make rw mount from view-only snapshot")
+				}
+
 				mounts, err = s.Mounts(ctx, target)
+				if err != nil {
+					return err
+				}
+
+				if info.Kind == snapshots.KindActive && !rw {
+					mounts[len(mounts)-1].Options = append(mounts[len(mounts)-1].Options, "ro")
```

Comment:
Just curious, all (including futural) snapshotters are guaranteed to provide readonly mount by this change (i.e. adding `ro` to the last element of `[]mount.Mount`)? Because the implementation detail (e.g. the backing filesystem of each `[]mount.Mount`) is completely hidden behind `Mounts()` API.

Reply:
Given the interface of the `Mount` type, I think it is fair to explicitly define some common cross platform options that should be supported such as `"ro"`.  I can't find any spots in the code today though where we are blindly adding the options to an undetermined mount though. I would at least call this out in a comment or update this to error in the same way the `rw` check is. Since it is `ctr` I think it is fine to try it out, but don't want that pattern done in other parts of containerd without a some more checking on how future-proof the approach is.

Reply:
Oh you're right this is not a linux specific file.
We could also, in a a platform specific file, add a new mount entry that remounts with ro?

Reply:
SGTM